opentelemetry: create a new module (#4361)
caddyserver/caddy: modules/caddyhttp/opentelemetry/README.md

# OpenTelemetry module

This module provides integration with OpenTelemetry tracing facilities. It is implemented
as `caddyhttp.MiddlewareHandler` and can be chained into a list of other handlers.

When enabled, it propagates an existing tracing context or will init a new one otherwise.

It is based on `https://github.com/open-telemetry/opentelemetry-go`.

## Configuration

### Environment variables

> It can be configured using environment variables defined
> by [spec](https://github.com/open-telemetry/opentelemetry-specification).

Reviewer: Can we link directly to the environment variables spec?

```suggestion
It can be configured using the environment variables defined
by the [OpenTelemetry Environment Variable Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md).
```

Author: Perfectly makes sense to me. Good point! 👍

IMPORTANT: Please, consider the version
of [https://github.com/open-telemetry/opentelemetry-go](https://github.com/open-telemetry/opentelemetry-go). Some parts

> of the specification may be not implemented yet.

Reviewer: The metrics and logs parts of the OTel spec are non-stable, but Tracing is stable so I'm not sure if this statement is accurate. Either way, it adds a bit of confusion that I'm not sure is necessary in this context.

Author: The reason why I put here this statement is this ticket: https://github.com/open-telemetry/opentelemetry-go/issues/1698. 

I just wanted to emphasize that customers should not expect that, for example, variable OTEL_TRACES_SAMPLER will change the behavior of the tracing.

What do you think, does it make sense to add this clarification or it is better to remove this line?

> Required by module environment variables are: OTEL_EXPORTER_OTLP_PROTOCOL/OTEL_EXPORTER_OTLP_TRACES_PROTOCOL and
> OTEL_PROPAGATORS.

Reviewer: How about:

```suggestion
These environment variables are required by this module:
- `OTEL_EXPORTER_OTLP_PROTOCOL` or `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL`
- `OTEL_PROPAGATORS`
```

(I'm not sure exactly what you mean though - maybe adjust this)

Author: You understood it absolutely correctly 😃 . And I think the proposed by you change sounds clearer. 👍

If neither OTEL_EXPORTER_OTLP_INSECURE nor OTEL_EXPORTER_OTLP_SPAN_INSECURE is provided, then:

1. If OTEL_EXPORTER_OTLP_CERTIFICATE or OTEL_EXPORTER_OTLP_TRACES_CERTIFICATE are specified they will be used for TLS.
2. Else if both OTEL_EXPORTER_OTLP_CERTIFICATE and OTEL_EXPORTER_OTLP_TRACES_CERTIFICATE are not specified, then default
   TLS with the default `tls.Config` config will be used for an exporter.

For the exporter configuration details, please
see [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/a4440931b522c7351b0485ff4899f786b4ff4459/specification/protocol/exporter.md)
.

Example:

```bash
export OTEL_EXPORTER_OTLP_HEADERS="myAuthHeader=myToken"
export OTEL_EXPORTER_OTLP_PROTOCOL=grpc
export OTEL_PROPAGATORS=tracecontext,baggage
export OTEL_EXPORTER_OTLP_SPAN_INSECURE=false
export OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=localhost:12345
```

### Caddy file configuration

Here is an example of **Caddyfile**:

```

> handle /myHanlder {

Reviewer: ```suggestion
handle /myHandler {
```

Author: Good catch 👍

# OpenTelemetry module

This module provides integration with OpenTelemetry tracing facilities. It is implemented

> as `caddyhttp.MiddlewareHandler` and can be chained into a list of other handlers.

Reviewer 2: ```suggestion
as a `caddyhttp.MiddlewareHandler` and can be chained into a list of other handlers.
```
